src/cpp/KnowledgeItem.cpp

```diff
@@ -47,3 +47,17 @@ void KnowledgeItem::save_to(BaseStorageSaver &storage) {
 	storage.save_creation_time(creation_time());
 	storage.save_title(title());
 }
+
+double KnowledgeItem::match(SearchCriteria<std::string, std::time_t> *criteria) {

```

Comment:
Implementations in subclasses will be forced to duplicate this code?
They can't just call base class method, because it returns passed/total rate, and updating it with other fields will be hard?

Reply:
When I was implementing this part I thought that subclasses can calculate their result as the sum of theirs and the base class' result divided by 2. Or this method isn't good for us?

Reply:
My idea is that the base class and subclass are like two separate entities that are equally responsible for the match of the whole object. 
For example:
There are 4 checks and 1 pass for a base class, so for him, it will be a 0.25 match.
There are 2 checks and 2 passes for a subclass, so for him, it will be a 1 match.
As a result base class brings 0.25 / 2 = 0.125 and subclass 1 / 2 = 0.5.
And total match for this object will be 0.625
And yes, in this idea moving of fields will influence the result.

Reply:
Let's keep this implementation for now, see if it works